memory-serve/build.rs

```diff
@@ -0,0 +1,256 @@
+                log!("skipping file {route}, could not determine file extension");
+                return None;
+            };
+
+            // do not load assets into the binary in debug / development mode
+            if cfg!(debug_assertions) {
+                log!("including {route} (dynamically)");
+
+                return Some(Asset {
+                    route,
+                    path: path.to_owned(),
+                    content_type,
+                    etag: Default::default(),
+                    compressed_bytes: None,
+                });
+            }
+
+            let Ok(bytes) = std::fs::read(entry.path()) else {
+                log!("skipping file {route}: file is not readable");
+                return None;
+            };
+
+            let etag: String = sha256::digest(&bytes);
+            let original_size = bytes.len();
+            let is_compress_type = COMPRESS_TYPES.contains(&content_type.as_str());
+            let brotli_bytes = if is_compress_type {
+                compress_brotli(&bytes)
+            } else {
+                None
+            };
+
+            let mut asset = Asset {
+                route: route.clone(),
+                path: path.to_owned(),
+                content_type,
+                etag,
+                compressed_bytes: None,
+            };
+
+            if is_compress_type {
+                match brotli_bytes {
+                    Some(brotli_bytes) if brotli_bytes.len() >= original_size => {
+                        log!("including {route} {original_size} bytes (compression unnecessary)");
+                    }
+                    Some(brotli_bytes) => {
+                        log!(
+                            "including {route} {original_size} -> {} bytes (compressed)",
+                            brotli_bytes.len()
+                        );
+
+                        asset.compressed_bytes = Some(brotli_bytes);
+                    }
+                    None => {
+                        log!("including {route} {original_size} bytes (compression failed)");
+                    }
+                }
+            } else {
+                log!("including {route} {original_size} bytes");
+            }
+
+            Some(asset)
+        })
+        .collect();
+
+    assets.sort();
+
+    assets
+}
+
+const ASSET_FILE: &str = "memory_serve_assets.rs";
+
+fn main() {
+    let out_dir: String = std::env::var("OUT_DIR").expect("OUT_DIR environment variable not set.");
+    let target = Path::new(&out_dir).join(ASSET_FILE);
+
+    let Ok(asset_dir) = std::env::var("ASSET_DIR") else {
+        log!("Please specify the `ASSET_DIR` environment variable.");
+
+        std::fs::write(target, "&[]").expect("Unable to write memory-serve asset file.");
+
+        println!("cargo::rerun-if-env-changed=ASSET_DIR");
+        return;
+    };
+
+    let path = Path::new(&asset_dir);
+    let path = path
```

Comment:
Unfortunately, the paths here are relative to the crate itself, instead of the crate implementing it. Is there a way so we can use paths relative to the implementing crate or relative to where the build script is called from?

Reply:
#16 addresses this